File: TDV - TP1/llenar_maquinas_BT.cpp
#include <iostream>
#include <vector>

using namespace std;

int max(vector<int> v){
    if(v.size() == 0){
        return 0;
    }
    int rv = v[0];
    for (int i = 0; i < v.size(); i++){
        if (v[i] > rv){
            rv = v[i];
        }
    }
    return rv;
}

bool vale_la_pena(vector<int> s, vector<int> beneficios, int & max){
    int benef_tot = 0;
    for(int i = 0; i < s.size(); i++){
        if (s[i] != 0){
            benef_tot += beneficios[i];
        }
    }
    for (int i = s.size(); i < beneficios.size(); i++){
        benef_tot += beneficios[i];
    }
    return benef_tot > max;
}

int llenar_mochila_BT_opt(vector<int> beneficios, vector<int> pesos_i, vector<int> pesos_m, vector<int> s, int & maximo){
    if (s.size() == beneficios.size()){
        if (pesos_ok(s, pesos_m, pesos_i)){
            int b = suma_beneficios(s, beneficios);
            if (b > maximo){
                maximo = b;
            }
            return b;
        }
        else{
            return 0;
        }
    }
    else{
        vector<int> valores;
        for (int i = 0; i < pesos_m.size()+1; i++){
            if (vale_la_pena(s, beneficios, maximo)){
                vector<int> s_aux = s;
                s_aux.push_back(i);
                valores.push_back(llenar_mochila_BT_opt(beneficios, pesos_i, pesos_m, s_aux, maximo));
            }
        }
        return max(valores);
    }
}

File: TDV - TP1/llenar_maquinas_BT.py
def pesos_ok(s, pesos_m, pesos_i):
    solucs = dict()
    for i in range(len(s)):
        if s[i] in solucs:
            solucs[s[i]] += pesos_i[i]
        else:
            solucs[s[i]] = pesos_i[i]
    solucs.pop(0, None)
    for clave in solucs:
        if solucs[clave] > pesos_m[clave-1]:
            return False
    return True

def suma_beneficios(s, beneficios):
    rv = 0
    for i in range(len(s)):
        if s[i] != 0:
            rv += beneficios[i]
    return rv

def peso_ok(maquina, pesos_i, s, capacidad):
    for i in range(len(s)):
        if s[i] == maquina:
            capacidad -= pesos_i[i]
    return capacidad >= pesos_i[len(s)]
a = 0
def llenar_maquinas_BT_fact(beneficios, pesos_i, pesos_m, s):
    global a
    a+=1
    if len(s) == len(beneficios):
        return suma_beneficios(s, beneficios)
    else: 
        valores = []
        for i in range(len(pesos_m)+1):
            if peso_ok(i, pesos_i, s, pesos_m[i-1]):
                s_aux = s[:] + [i]
                valores.append(llenar_maquinas_BT_fact(beneficios, pesos_i, pesos_m, s_aux))         
        return max(valores) 
 
    
maquinas = [10, 10]
pesos = [5, 5, 6, 4, 1]
beneficios = [5, 5, 5, 5, 1]

print(llenar_maquinas_BT_fact(beneficios, pesos, maquinas, []))
print("a = ", a)
